Expose reason for notification dismissal through SetInt()

lanurmi opened the pull request:
On all platforms, except macOS, the reason why a notification was dismissed is known. Specifically, it can be useful to distinguish whether the notification closed because of a timeout, or a click by the user.

This is left undocumented on purpose for now. The values passed are platform-dependent, and this is not the most elegant API.

lanurmi committed:
Expose reason for notification dismissal through SetInt()

On all platforms, except macOS, the reason why a notification was
dismissed is known. Specifically, it can be useful to distinguish whether
the notification closed because of a timeout, or a click by the user.

vadz reviewed (commented):
Thanks, this can definitely be useful, but we need to name and document these constants instead of just using `2` in the code.

I guess a (scoped) enum could be defined in `wxNotificationMessage` itself?

lanurmi commented:
> Thanks, this can definitely be useful, but we need to name and document these constants instead of just using `2` in the code.

We can do that, yes. There is just the MSW value `ToastDismissalReason_ApplicationHidden` that does not correspond to anything that the Gtk implementation would currently ever return. But at a closer look it does correspond to reason 3 in the Desktop Notifications Specification, so there is no fundamental mismatch of reason meanings across platforms.

> I guess a (scoped) enum could be defined in `wxNotificationMessage` itself?

I'll do that.

lanurmi commented:
Done with modifications. Since `SetInt()` is not used for anything yet, might as well use that instead of `SetExtraLong()`. The documentation assumes this feature will be available since 3.2.5.

All commits should probably be squashed into one.

vadz commented on src/generic/notifmsgg.cpp:
```diff
@@ -265,6 +265,7 @@ void wxNotificationMessageWindow::Set(int timeout)
 void wxNotificationMessageWindow::OnClose(wxCloseEvent& WXUNUSED(event))
 {
     wxCommandEvent evt(wxEVT_NOTIFICATION_MESSAGE_DISMISSED);
+    evt.SetExtraLong(0);
```
This should use the symbolic name now (even though it would require a cast).

vadz commented on src/generic/notifmsgg.cpp:
```diff
@@ -307,6 +308,7 @@ void wxNotificationMessageWindow::OnNotificationMouseLeave(wxMouseEvent& WXUNUSE
 void wxNotificationMessageWindow::OnCloseClicked(wxCommandEvent& WXUNUSED(event))
 {
     wxCommandEvent evt(wxEVT_NOTIFICATION_MESSAGE_DISMISSED);
+    evt.SetExtraLong(2);
```
Same as above.

lanurmi replied:
Looks like I forgot to modify under `generic/`.

vadz commented on src/gtk/notifmsg.cpp:
```diff
@@ -312,12 +312,28 @@ class wxLibNotifyMsgImpl : public wxNotificationMessageImpl
+            break;
+        }
+
         switch (closeReason)
```
Sorry, it doesn't make to have 2 `switch (closeReason)` to me in a row. This should probably be changed to `switch (reason)` now?

lanurmi replied:
It makes some sense, as the cases aren't the same in both. But yes, `switch (reason)` for the latter would be better.

vadz commented on interface/wx/notifmsg.h:
```diff
@@ -64,6 +67,15 @@ class wxNotificationMessage : public wxEvtHandler
         Timeout_Never = 0   ///< Notification will never time out.
     };
 
+    /// Reasons for dismissal, posted with wxEVT_NOTIFICATION_MESSAGE_DISMISSED events.
```
Although trivial, we probably still want to do this for consistency:

```suggestion
    /**
      Reasons for dismissal, posted with wxEVT_NOTIFICATION_MESSAGE_DISMISSED events.
      
      @since 3.2.5
    */
```

vadz commented on include/wx/notifmsg.h:
```diff
@@ -81,6 +81,15 @@ class WXDLLIMPEXP_CORE wxNotificationMessageBase : public wxEvtHandler
+    enum class DismissalReason
+    {
+        Unknown,
+        DismissedByUser,
+        DismissedByApp,
```
Using `DismissalReason::DismissedByUser` in the code looks a bit heavy, maybe we should call this just `ByUser`/`ByApp`?

lanurmi replied:
Sounds good.

vadz reviewed (commented):
Thanks for the update!

Is it important to have this in 3.2.5? We don't require C++11 there, so we can't use a scoped enum and would have to replace it with a weak one.

lanurmi commented:
> Is it important to have this in 3.2.5? We don't require C++11 there, so we can't use a scoped enum and would have to replace it with a weak one.

Not terribly important for me personally, but would be nice to get such simple new features into Linux distributions and such within this decade. There is also the API-forward-compatible approach of using just numerical values in 3.2.x, matching the values of the scoped enum. And only introduce the the scoped enum in 3.3.

lanurmi commented:
I have now altered the since versions to 3.3.0, as backporting to 3.2.x is not certain yet.

vadz commented on src/gtk/notifmsg.cpp:
```diff
@@ -312,15 +312,33 @@ class wxLibNotifyMsgImpl : public wxNotificationMessageImpl
                 ProcessNotificationEvent(evt);
                 break;
             }
+            default:
```
Very minor, but could you please explicitly list the 2 other cases here instead of using `default:`? This would ensure that we get a warning from the compiler if we ever add another enum element and forget to handle it here.

vadz commented on include/wx/notifmsg.h:
```diff
@@ -81,6 +81,15 @@ class WXDLLIMPEXP_CORE wxNotificationMessageBase : public wxEvtHandler
     };
 
+    // reasons for dismissal, posted with wxEVT_NOTIFICATION_MESSAGE_DISMISSED events
+    enum class DismissalReason
```
If we want to backport this to 3.2, we'd need to make this plain (unscoped) enum, e.g.

```cpp
enum DismissalReason
{
  Reason_Unknown,
  ...
};
```

vadz reviewed (commented):
Thanks for the update!

I have just one question remaining before merging this into master: are we supposed to get notifications for `ByApp` reason or not? It looks to me (but I didn't test it) that GTK version avoids sending those, but MSW one seems to do it. Shouldn't the latter one avoid them too?

And, as already mentioned, if we/you want to have this in 3.2, we should avoid using C++11 scoped enums. But it's perfectly fine to have this enum for 3.3 only, of course.

lanurmi commented:
> Thanks for the update!
> 
> I have just one question remaining before merging this into master: are we supposed to get notifications for `ByApp` reason or not? It looks to me (but I didn't test it) that GTK version avoids sending those, but MSW one seems to do it. Shouldn't the latter one avoid them too?

I have no definite answer to this. I have been wondering why the GTK version avoids sending them (maybe there is a reason, but it is undocumented). So I figured changing existing behavior is out of scope for this PR, and chose to let both GTK and MSW keep doing/not doing what they were before.

I have pushed the latest requested change to `default:`, along with some indentation adjustments and two `auto` keywords.

> And, as already mentioned, if we/you want to have this in 3.2, we should avoid using C++11 scoped enums. But it's perfectly fine to have this enum for 3.3 only, of course.

I wouldn't want to be limited by C++98 in 3.3, so let us stick to the scoped enum. The backporting options would be:
1. Just not to introduce any `enum` in 3.2.
2. Or maybe:
```c++
class wxNotificationMessage {
   ...
   struct DismissalReason { enum { Unknown, ByUser, ByApp, TimedOut } };
```
This would be _mostly_ source code compatible with how a scoped enum is used. I'm not 100% sure if adding a `struct` into a `class` changes ABI, even though there is no instance of the `struct`.

vadz commented:
> > Thanks for the update!

(And again)

> > I have just one question remaining before merging this into master: are we supposed to get notifications for `ByApp` reason or not? It looks to me (but I didn't test it) that GTK version avoids sending those, but MSW one seems to do it. Shouldn't the latter one avoid them too?
> 
> I have no definite answer to this. I have been wondering why the GTK version avoids sending them (maybe there is a reason, but it is undocumented). So I figured changing existing behavior is out of scope for this PR, and chose to let both GTK and MSW keep doing/not doing what they were before.

I'd like to at least decide if we're going to standardize on the same behaviour in the future (as I hope) and, if so, whether we even need `ByApp` dismissal reason: if dismissing the notification by the application itself shouldn't produce any notification, it doesn't seem useful to introduce it.

> I wouldn't want to be limited by C++98 in 3.3, so let us stick to the scoped enum. The backporting options would be:
> 
>     1. Just not to introduce any `enum` in 3.2.
> 
>     2. Or maybe:
> 
> 
> ```c++
> class wxNotificationMessage {
>    ...
>    struct DismissalReason { enum { Unknown, ByUser, ByApp, TimedOut } };
> ```
> 
> This would be _mostly_ source code compatible with how a scoped enum is used.

Yes, we could do this...

> I'm not 100% sure if adding a `struct` into a `class` changes ABI, even though there is no instance of the `struct`.

Adding things is, in general, ABI-compatible.

lanurmi committed:
Always post notification dismissal event on Gtk, regardless of reason

For an unknown reason, the Gtk implementation only used to post a
wxEVT_NOTIFICATION_MESSAGE_DISMISSED event if the associacted reason
code was either 1 or 2. To be consistent with with MSW, post the
event regardless of what the reason was, potentially enabling the
event for reasons ByApp and Unknown.

lanurmi commented:
> > > I have just one question remaining before merging this into master: are we supposed to get notifications for `ByApp` reason or not? 
> I'd like to at least decide if we're going to standardize on the same behaviour in the future (as I hope) and, if so, whether we even need `ByApp` dismissal reason: if dismissing the notification by the application itself shouldn't produce any notification, it doesn't seem useful to introduce it.

I have added a separate commit that changes GTK's behavior so that it sends the event with all reason values, not just the two.

vadz commented:
> I have added a separate commit that changes GTK's behavior so that it sends the event with all reason values, not just the two.

Thanks! I'll merge this soon.
